In [1]:
import pandas as pd

from io import StringIO

# ====csv file create====

## .to_csv()=> csv file create kore

## ==========Store data into a csv file======

## .to_csv()=dataframe ke csv te convert kore,csv file e store kore

## ==> (dataframe name).to_csv(directiory location where want to store/jai file e store korte chai) #header =true ..means write the header #header =false ..means dont write the header

In [69]:
a=pd.DataFrame([[12,22,32,50],[12,29,32,60],[10,50,60,90]],index=["row1","row2","row3"],
               columns=["col1","col2","col3","col4"])

print(a,"\n","\n")

a.to_csv("demo1.csv")

print(a)

      col1  col2  col3  col4
row1    12    22    32    50
row2    12    29    32    60
row3    10    50    60    90 
 

      col1  col2  col3  col4
row1    12    22    32    50
row2    12    29    32    60
row3    10    50    60    90


In [70]:
#create a dataframe and store to csv file

import pandas as pd
import numpy as np
d=pd.DataFrame(np.arange(1,16).reshape(3,5),index=["row1","row2","row3"],
               columns=["col1","col2","col3","col4","col5"])

print(d)

d.to_csv("sweety.csv")

      col1  col2  col3  col4  col5
row1     1     2     3     4     5
row2     6     7     8     9    10
row3    11    12    13    14    15


In [72]:
x=d.reindex(["r1","r2","r3"])
print(x)

    col1  col2  col3  col4  col5
r1   NaN   NaN   NaN   NaN   NaN
r2   NaN   NaN   NaN   NaN   NaN
r3   NaN   NaN   NaN   NaN   NaN


## Read CSV Files
 
## >=A simple way to store big data sets is to use CSV files (comma separated files).

## >=CSV files contains plain text and is a well know format that can be read by everyone including Pandas.

# >>>>>> Load Data <<<<<<<<

### ===Load data from csv file====

## df=pd.read_csv("file name.csv")


### working Directory = the directory in my machine where the python software is loaded

### if the file is in that path it will look in that path

### if it in other path we have to mention the full path..c:\myfolder....

#### >==it will read all data there in the csv file and put them into dataframe

#### >==1st row in csv file will become the 0th row in the dataframe index=0

#### >==but some times in the 1st row of the csv file we have the column names/headers

#### >==so if file contains header we have to mention (header=0) means 0th index/row is actually my header information

#### >==if no header is present in my file i can skip (header=0) part...dnt need to mention

In [73]:
a=pd.read_csv("onewaydata.csv")
print(a)

   60  50  48  17
0  67  52  49  67
1  42  43  50  54
2  67  67  55  67
3  56  67  56  68
4  62  59  61  65
5  64  67  61  65
6  59  64  60  56
7  72  63  59  60
8  71  65  64  65


In [74]:
a=pd.read_csv("onewaydata.csv",header=0)
#1st row in csv file will become column names
print(a)

   60  50  48  17
0  67  52  49  67
1  42  43  50  54
2  67  67  55  67
3  56  67  56  68
4  62  59  61  65
5  64  67  61  65
6  59  64  60  56
7  72  63  59  60
8  71  65  64  65


In [75]:
b=pd.read_csv("onewaydata.csv",header=None)
print(b)

    0   1   2   3
0  60  50  48  17
1  67  52  49  67
2  42  43  50  54
3  67  67  55  67
4  56  67  56  68
5  62  59  61  65
6  64  67  61  65
7  59  64  60  56
8  72  63  59  60
9  71  65  64  65


In [ ]:
# a excel file can have multiple sheet so we need to mention sheet name
a=pd.read_csv("onewaydata.csv",sheet_name="sheet1")
print(a)

# ======= Read a csv File=======

### .read_csv() => csv file ke read kore

In [76]:
import pandas as pd
b=pd.read_csv("demo.csv")
print(b)

  \tcol1\tcol2\tcol3\tcol4\tcol5\tcol6\tcol7\tcol8\tcol9\tcol10
0           row1\t1\t\t21\t31\t41\t51\t61\t71\t81\tA           
1         row2\t200\t\t22\t32\t42\t52\t62\t72\t82\tB           
2           row3\t3\t13\t23\t33\t43\t53\t63\t73\t\tC           
3          row4\t0\t14\t24\t0\t44\t54\t64\t74\t84\tD           
4         row5\t5\t15\t25\t35\t45\t55\t65\t75\t85\tE           
5         row6\t6\t16\t26\t36\t46\t56\t66\t76\t86\tF           
6         row7\t7\t17\t27\t37\t47\t57\t67\t77\t87\tG           
7         row8\t8\t18\t28\t38\t48\t58\t68\t78\t88\tH           
8             row9\t9\t19\t\t39\t49\t\t69\t79\t89\tI           
9       row10\t10\t20\t30\t40\t50\t60\t70\t80\t90\tJ           


### csv file e comma dia value separate kora tykle tyle dataframe e alada alada kore value ta nybe jodi onno ki6u dia separate kora tyke file e tyle syta read korar somay define kore dite hobe , use korle value gulo separate hoy by default but akhane \n dia new line and \t= tab use kore space dia6e akhane default separator nai so amke separator define kore dite hobe sep='\t'

In [77]:
b=pd.read_csv("demo.csv",sep="\t")
print(b)

  Unnamed: 0  col1  col2  col3  col4  col5  col6  col7  col8  col9 col10
0       row1     1   NaN  21.0    31    41  51.0    61    71  81.0     A
1       row2   200   NaN  22.0    32    42  52.0    62    72  82.0     B
2       row3     3  13.0  23.0    33    43  53.0    63    73   NaN     C
3       row4     0  14.0  24.0     0    44  54.0    64    74  84.0     D
4       row5     5  15.0  25.0    35    45  55.0    65    75  85.0     E
5       row6     6  16.0  26.0    36    46  56.0    66    76  86.0     F
6       row7     7  17.0  27.0    37    47  57.0    67    77  87.0     G
7       row8     8  18.0  28.0    38    48  58.0    68    78  88.0     H
8       row9     9  19.0   NaN    39    49   NaN    69    79  89.0     I
9      row10    10  20.0  30.0    40    50  60.0    70    80  90.0     J


In [78]:
b=pd.read_csv("demo.csv",sep="\t",header=0)
print(b)

  Unnamed: 0  col1  col2  col3  col4  col5  col6  col7  col8  col9 col10
0       row1     1   NaN  21.0    31    41  51.0    61    71  81.0     A
1       row2   200   NaN  22.0    32    42  52.0    62    72  82.0     B
2       row3     3  13.0  23.0    33    43  53.0    63    73   NaN     C
3       row4     0  14.0  24.0     0    44  54.0    64    74  84.0     D
4       row5     5  15.0  25.0    35    45  55.0    65    75  85.0     E
5       row6     6  16.0  26.0    36    46  56.0    66    76  86.0     F
6       row7     7  17.0  27.0    37    47  57.0    67    77  87.0     G
7       row8     8  18.0  28.0    38    48  58.0    68    78  88.0     H
8       row9     9  19.0   NaN    39    49   NaN    69    79  89.0     I
9      row10    10  20.0  30.0    40    50  60.0    70    80  90.0     J


# ==========head() ================

### => by default csv file tyke 1st 5 ta row show kore/tule ane

### => kono value pass korale sykota dybe

In [79]:
#=====head bydefalut 5 ta value day

b=pd.read_csv("demo.csv",sep="\t")
print(b.head())

  Unnamed: 0  col1  col2  col3  col4  col5  col6  col7  col8  col9 col10
0       row1     1   NaN  21.0    31    41  51.0    61    71  81.0     A
1       row2   200   NaN  22.0    32    42  52.0    62    72  82.0     B
2       row3     3  13.0  23.0    33    43  53.0    63    73   NaN     C
3       row4     0  14.0  24.0     0    44  54.0    64    74  84.0     D
4       row5     5  15.0  25.0    35    45  55.0    65    75  85.0     E


In [80]:
b=pd.read_csv("demo.csv",sep="\t")
print(b.head(8))

  Unnamed: 0  col1  col2  col3  col4  col5  col6  col7  col8  col9 col10
0       row1     1   NaN  21.0    31    41  51.0    61    71  81.0     A
1       row2   200   NaN  22.0    32    42  52.0    62    72  82.0     B
2       row3     3  13.0  23.0    33    43  53.0    63    73   NaN     C
3       row4     0  14.0  24.0     0    44  54.0    64    74  84.0     D
4       row5     5  15.0  25.0    35    45  55.0    65    75  85.0     E
5       row6     6  16.0  26.0    36    46  56.0    66    76  86.0     F
6       row7     7  17.0  27.0    37    47  57.0    67    77  87.0     G
7       row8     8  18.0  28.0    38    48  58.0    68    78  88.0     H


# =================info()================

### => csv file e kotogulo row,column a6e
### =>kon column e ki type data a6e  sob return kore

In [81]:
#===================info()===

b=pd.read_csv("demo.csv",sep="\t")
print(b.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  10 non-null     object 
 1   col1        10 non-null     int64  
 2   col2        8 non-null      float64
 3   col3        9 non-null      float64
 4   col4        10 non-null     int64  
 5   col5        10 non-null     int64  
 6   col6        9 non-null      float64
 7   col7        10 non-null     int64  
 8   col8        10 non-null     int64  
 9   col9        9 non-null      float64
 10  col10       10 non-null     object 
dtypes: float64(4), int64(5), object(2)
memory usage: 1008.0+ bytes
None


# =================describe()=============

### ==> count kon column e kota value a6e =>mean
### =>standard deviation 
### =>minimum value 
### =>kono column e object type r data tykle syta return kore na...syta count kore na 
### =>parcentile show kore 25,50,75 parcentile 
### =>maximum value return kore

In [82]:
#============describe()

b=pd.read_csv("demo.csv",sep="\t")
print(b.describe())

             col1      col2       col3      col4      col5       col6  \
count   10.000000   8.00000   9.000000  10.00000  10.00000   9.000000   
mean    24.900000  16.50000  25.111111  32.10000  45.50000  55.111111   
std     61.612499   2.44949   2.934469  11.66619   3.02765   2.934469   
min      0.000000  13.00000  21.000000   0.00000  41.00000  51.000000   
25%      3.500000  14.75000  23.000000  32.25000  43.25000  53.000000   
50%      6.500000  16.50000  25.000000  35.50000  45.50000  55.000000   
75%      8.750000  18.25000  27.000000  37.75000  47.75000  57.000000   
max    200.000000  20.00000  30.000000  40.00000  50.00000  60.000000   

           col7      col8       col9  
count  10.00000  10.00000   9.000000  
mean   65.50000  75.50000  85.777778  
std     3.02765   3.02765   3.073181  
min    61.00000  71.00000  81.000000  
25%    63.25000  73.25000  84.000000  
50%    65.50000  75.50000  86.000000  
75%    67.75000  77.75000  88.000000  
max    70.00000  80.00000  90.

In [83]:
b=pd.read_csv("demo.csv",sep="\t")
print(b.min())

Unnamed: 0    row1
col1             0
col2          13.0
col3          21.0
col4             0
col5            41
col6          51.0
col7            61
col8            71
col9          81.0
col10            A
dtype: object


# ==============value_counts()==============

### >>count kon value kotobar a6e

In [84]:
b=pd.read_csv("demo.csv",sep="\t")
print(b["col1"].value_counts())

0      1
1      1
3      1
10     1
5      1
6      1
7      1
200    1
9      1
8      1
Name: col1, dtype: int64


# ========Accessing the elements=======

In [85]:
b=pd.read_csv("demo.csv",sep="\t")
print(b)

  Unnamed: 0  col1  col2  col3  col4  col5  col6  col7  col8  col9 col10
0       row1     1   NaN  21.0    31    41  51.0    61    71  81.0     A
1       row2   200   NaN  22.0    32    42  52.0    62    72  82.0     B
2       row3     3  13.0  23.0    33    43  53.0    63    73   NaN     C
3       row4     0  14.0  24.0     0    44  54.0    64    74  84.0     D
4       row5     5  15.0  25.0    35    45  55.0    65    75  85.0     E
5       row6     6  16.0  26.0    36    46  56.0    66    76  86.0     F
6       row7     7  17.0  27.0    37    47  57.0    67    77  87.0     G
7       row8     8  18.0  28.0    38    48  58.0    68    78  88.0     H
8       row9     9  19.0   NaN    39    49   NaN    69    79  89.0     I
9      row10    10  20.0  30.0    40    50  60.0    70    80  90.0     J


In [86]:
print(b["col1"])

0      1
1    200
2      3
3      0
4      5
5      6
6      7
7      8
8      9
9     10
Name: col1, dtype: int64


In [87]:
print(b.loc[0])

Unnamed: 0    row1
col1             1
col2           NaN
col3          21.0
col4            31
col5            41
col6          51.0
col7            61
col8            71
col9          81.0
col10            A
Name: 0, dtype: object


In [88]:
print(type(b.loc[0]))

<class 'pandas.core.series.Series'>


In [89]:
print(type(b))

<class 'pandas.core.frame.DataFrame'>


### Take the elements from the Column2

In [90]:
print(b.iloc[:,1:])

   col1  col2  col3  col4  col5  col6  col7  col8  col9 col10
0     1   NaN  21.0    31    41  51.0    61    71  81.0     A
1   200   NaN  22.0    32    42  52.0    62    72  82.0     B
2     3  13.0  23.0    33    43  53.0    63    73   NaN     C
3     0  14.0  24.0     0    44  54.0    64    74  84.0     D
4     5  15.0  25.0    35    45  55.0    65    75  85.0     E
5     6  16.0  26.0    36    46  56.0    66    76  86.0     F
6     7  17.0  27.0    37    47  57.0    67    77  87.0     G
7     8  18.0  28.0    38    48  58.0    68    78  88.0     H
8     9  19.0   NaN    39    49   NaN    69    79  89.0     I
9    10  20.0  30.0    40    50  60.0    70    80  90.0     J


# convert Dataframes into array

In [91]:
b.iloc[:,1:].values

array([[1, nan, 21.0, 31, 41, 51.0, 61, 71, 81.0, 'A'],
       [200, nan, 22.0, 32, 42, 52.0, 62, 72, 82.0, 'B'],
       [3, 13.0, 23.0, 33, 43, 53.0, 63, 73, nan, 'C'],
       [0, 14.0, 24.0, 0, 44, 54.0, 64, 74, 84.0, 'D'],
       [5, 15.0, 25.0, 35, 45, 55.0, 65, 75, 85.0, 'E'],
       [6, 16.0, 26.0, 36, 46, 56.0, 66, 76, 86.0, 'F'],
       [7, 17.0, 27.0, 37, 47, 57.0, 67, 77, 87.0, 'G'],
       [8, 18.0, 28.0, 38, 48, 58.0, 68, 78, 88.0, 'H'],
       [9, 19.0, nan, 39, 49, nan, 69, 79, 89.0, 'I'],
       [10, 20.0, 30.0, 40, 50, 60.0, 70, 80, 90.0, 'J']], dtype=object)

In [98]:
b.iloc[0].values

array(['row1', 1, nan, 21.0, 31, 41, 51.0, 61, 71, 81.0, 'A'],
      dtype=object)

In [94]:
#corelation

b.corr()

col1      col2      col3      col4      col5      col6      col7  \
col1  1.000000  0.920390 -0.355190  0.037424 -0.360064 -0.355190 -0.360064   
col2  0.920390  1.000000  1.000000  0.560244  1.000000  1.000000  1.000000   
col3 -0.355190  1.000000  1.000000  0.375397  1.000000  1.000000  1.000000   
col4  0.037424  0.560244  0.375397  1.000000  0.419956  0.375397  0.419956   
col5 -0.360064  1.000000  1.000000  0.419956  1.000000  1.000000  1.000000   
col6 -0.355190  1.000000  1.000000  0.375397  1.000000  1.000000  1.000000   
col7 -0.360064  1.000000  1.000000  0.419956  1.000000  1.000000  1.000000   
col8 -0.360064  1.000000  1.000000  0.419956  1.000000  1.000000  1.000000   
col9 -0.417378  1.000000  1.000000  0.447214  1.000000  1.000000  1.000000   

          col8      col9  
col1 -0.360064 -0.417378  
col2  1.000000  1.000000  
col3  1.000000  1.000000  
col4  0.419956  0.447214  
col5  1.000000  1.000000  
col6  1.000000  1.000000  
col7  1.000000  1.000000  
col8  1.000000  1.000000  
col9  1.000000  1.000000

# =======================header=====================

### column name change

In [99]:
import pandas as pd

b=pd.read_csv("demo.csv",sep="\t")
print(b)

  Unnamed: 0  col1  col2  col3  col4  col5  col6  col7  col8  col9 col10
0       row1     1   NaN  21.0    31    41  51.0    61    71  81.0     A
1       row2   200   NaN  22.0    32    42  52.0    62    72  82.0     B
2       row3     3  13.0  23.0    33    43  53.0    63    73   NaN     C
3       row4     0  14.0  24.0     0    44  54.0    64    74  84.0     D
4       row5     5  15.0  25.0    35    45  55.0    65    75  85.0     E
5       row6     6  16.0  26.0    36    46  56.0    66    76  86.0     F
6       row7     7  17.0  27.0    37    47  57.0    67    77  87.0     G
7       row8     8  18.0  28.0    38    48  58.0    68    78  88.0     H
8       row9     9  19.0   NaN    39    49   NaN    69    79  89.0     I
9      row10    10  20.0  30.0    40    50  60.0    70    80  90.0     J


In [100]:
#row 1 become header,
#,row1's value become column name now if we want to access column element we have to access with row1 element name

b=pd.read_csv("demo.csv",sep="\t",header=1)
print(b)

    row1    1  Unnamed: 2    21  31  41    51  61  71    81  A
0   row2  200         NaN  22.0  32  42  52.0  62  72  82.0  B
1   row3    3        13.0  23.0  33  43  53.0  63  73   NaN  C
2   row4    0        14.0  24.0   0  44  54.0  64  74  84.0  D
3   row5    5        15.0  25.0  35  45  55.0  65  75  85.0  E
4   row6    6        16.0  26.0  36  46  56.0  66  76  86.0  F
5   row7    7        17.0  27.0  37  47  57.0  67  77  87.0  G
6   row8    8        18.0  28.0  38  48  58.0  68  78  88.0  H
7   row9    9        19.0   NaN  39  49   NaN  69  79  89.0  I
8  row10   10        20.0  30.0  40  50  60.0  70  80  90.0  J


# ============== tail() method =========

In [11]:
x=("d,e,f \n"
"1,a,b, \n"
"2,x,y, \n"
"3,m,n,")

df=pd.read_csv(StringIO(x))

print(df)


#here value is 4 for every row row 0 te 1,a,b and comma a6e but value nai
#bydefault 1st column ta integer bole sytake index no hisabe nia6e

   d  e   f 
1  a  b     
2  x  y     
3  m  n  NaN


## don't want 1st column to be index no

In [13]:
x=("d,e,f \n"
"1,a,b, \n"
"2,x,y, \n"
"3,m,n,")

df=pd.read_csv(StringIO(x),index_col=False)

print(df)

   d  e f 
0  1  a  b
1  2  x  y
2  3  m  n


# =========Escape character========

### to remove any character from dataframe

In [6]:
data='a,b\n"hello,\\"How are you\\",today is wednesday",2021 '

df=pd.read_csv(StringIO(data))

print(df)


#a,b column name

                                        a     b
hello,\How are you\"  today is wednesday"  2021


In [7]:
data='a,b\n"hello,\\"How are you\\",today is wednesday",2021 '

df=pd.read_csv(StringIO(data),escapechar="\\")

print(df)


                                        a     b
0  hello,"How are you",today is wednesday  2021


# ========Json Data read==========

.....api return json data

In [ ]:
#from url


data="..url.."
df=pd.read_json(data)
print(df)

In [3]:
#read json to csv

Data ='{"employee_name": "James", "email": "james@gmail.com", "job_profile": [{"title1":"Team Lead", "title2":"Sr. Developer"}]}'
df=pd.read_json(Data)

print(df)

  employee_name            email  \
0         James  james@gmail.com   

                                         job_profile  
0  {'title1': 'Team Lead', 'title2': 'Sr. Develop...  


# ===================HTML data Read============

## Reading HTML content

In [4]:
#wikipedia r currency page e ja ja table 6ilo sob read kore nilam


data="https://en.wikipedia.org/wiki/Currency"

df=pd.read_html(data)

print(df)

[    0                                                  1
0 NaN  Look up currency in Wiktionary, the free dicti...,                                           Numismatics
0                                            Currency
1                    Coins Banknotes Forgery List ISO
2                              Circulating currencies
3             Africa The Americas Europe Asia Oceania
4                                    Local currencies
5                     Company scrip LETS Time dollars
6             Fictional currenciesProposed currencies
7                                             History
8                               Historical currencies
9   Greek Roman China India Persian Tibetan Thai F...
10                                          Byzantine
11                                Medieval currencies
12                                         Production
13      Mint Designers Coining Milling Hammering Cast
14                                           Exonumia
15                 C

if there are lots of table we can fetch only 1 table by giving table index no

In [5]:
#only print particular table
#1st table e ja 6ilo only seta

data="https://en.wikipedia.org/wiki/Currency"

df=pd.read_html(data)

print(df[0])

    0                                                  1
0 NaN  Look up currency in Wiktionary, the free dicti...


In [6]:
data="https://en.wikipedia.org/wiki/Currency"

df=pd.read_html(data)

print(df[4])

             Rank              Currency ISO 4217 code(symbol)  \
0               1  United States dollar             USD (US$)   
1               2                  Euro               EUR (€)   
2               3          Japanese yen               JPY (¥)   
3               4        Pound sterling               GBP (£)   
4               5     Australian dollar              AUD (A$)   
5               6       Canadian dollar              CAD (C$)   
6               7           Swiss franc             CHF (CHF)   
7               8              Renminbi           CNY (元 / ¥)   
8               9      Hong Kong dollar             HKD (HK$)   
9              10    New Zealand dollar             NZD (NZ$)   
10             11         Swedish krona              SEK (kr)   
11             12      South Korean won               KRW (₩)   
12             13      Singapore dollar              SGD (S$)   
13             14       Norwegian krone              NOK (kr)   
14             15        

In [7]:
#currency column name ta je je table e tykbe jai protek ta table o read korbe

data="https://en.wikipedia.org/wiki/Currency"

df=pd.read_html(data,match="Currency")

print(df[4])

    Rank              Currency January 2018            Currency.1 August 2019
0    NaN                 World      100.00%                 World     100.00%
1    1.0  United States dollar       38.53%  United States dollar      42.52%
2    2.0                  Euro       32.75%                  Euro      32.06%
3    3.0        Pound sterling        7.22%        Pound sterling       6.21%
4    4.0          Japanese yen        2.80%          Japanese yen       3.61%
5    5.0              Renminbi        1.66%              Renminbi       2.22%
6    6.0       Canadian dollar        1.51%       Canadian dollar       1.76%
7    7.0           Swiss franc        1.42%     Australian dollar       1.57%
8    8.0     Australian dollar        1.38%      Hong Kong dollar       1.48%
9    9.0      Hong Kong dollar        1.32%             Thai baht       1.00%
10  10.0      Singapore dollar        1.01%      Singapore dollar       0.98%
11  11.0             Thai baht        0.95%           Swiss fran

In [8]:
#header change


data="https://en.wikipedia.org/wiki/Currency"

df=pd.read_html(data,match="Currency",header=1)

print(df[4])

  Economic theory Political economy Applied economics  \
0                                        Methodology    
1                                     Microeconomics    
2                                     Macroeconomics    
3                              Mathematicaleconomics    
4                                     Applied fields    
5               Schools (history)of economic thought    
6     Notable economistsand thinkerswithin economics    
7                         Internationalorganizations    
8  Category Index Lists Outline Publications Busi...    

  Economic theory Political economy Applied economics.1  
0  Economic model Economic systems Microfoundatio...     
1  Aggregation problem Budget set Consumer choice...     
2  Aggregate demand Balance of payments Business ...     
3  Contract theory Decision theory Econometrics G...     
4  Agricultural Business Demographic Development ...     
5  American (National) Ancient thought Anarchist ...     
6  François Quesnay Ada

# Reading EXcel Files

In [ ]:
data="Excel_Sample.xlsx/url"
df=pd.read_excel(data)
print(df)

In [ ]:
======Load data from excel file======

df=pd.read_excel("path",sheet_name="sheet1",header=0)

====write a csv===

df=pd.to_excel("path",sheet_name="sheet1",header=True)


# =================drop() =delete row===========

In [65]:
d=pd.read_csv("demo.csv",sep="\t")
print(d,'\n','\n')

d.drop(1,inplace=True)

print(d,'\n','\n')

  Unnamed: 0  col1  col2  col3  col4  col5  col6  col7  col8  col9 col10
0       row1     1   NaN  21.0    31    41  51.0    61    71  81.0     A
1       row2   200   NaN  22.0    32    42  52.0    62    72  82.0     B
2       row3     3  13.0  23.0    33    43  53.0    63    73   NaN     C
3       row4     0  14.0  24.0     0    44  54.0    64    74  84.0     D
4       row5     5  15.0  25.0    35    45  55.0    65    75  85.0     E
5       row6     6  16.0  26.0    36    46  56.0    66    76  86.0     F
6       row7     7  17.0  27.0    37    47  57.0    67    77  87.0     G
7       row8     8  18.0  28.0    38    48  58.0    68    78  88.0     H
8       row9     9  19.0   NaN    39    49   NaN    69    79  89.0     I
9      row10    10  20.0  30.0    40    50  60.0    70    80  90.0     J 
 

  Unnamed: 0  col1  col2  col3  col4  col5  col6  col7  col8  col9 col10
0       row1     1   NaN  21.0    31    41  51.0    61    71  81.0     A
2       row3     3  13.0  23.0    33    43  53.

In [63]:
d=pd.read_csv("demo.csv",sep="\t")
print(d,'\n','\n')

d.drop([4,5,6],inplace=True)

print(d,'\n','\n')

  Unnamed: 0  col1  col2  col3  col4  col5  col6  col7  col8  col9 col10
0       row1     1   NaN  21.0    31    41  51.0    61    71  81.0     A
1       row2   200   NaN  22.0    32    42  52.0    62    72  82.0     B
2       row3     3  13.0  23.0    33    43  53.0    63    73   NaN     C
3       row4     0  14.0  24.0     0    44  54.0    64    74  84.0     D
4       row5     5  15.0  25.0    35    45  55.0    65    75  85.0     E
5       row6     6  16.0  26.0    36    46  56.0    66    76  86.0     F
6       row7     7  17.0  27.0    37    47  57.0    67    77  87.0     G
7       row8     8  18.0  28.0    38    48  58.0    68    78  88.0     H
8       row9     9  19.0   NaN    39    49   NaN    69    79  89.0     I
9      row10    10  20.0  30.0    40    50  60.0    70    80  90.0     J 
 

  Unnamed: 0  col1  col2  col3  col4  col5  col6  col7  col8  col9 col10
0       row1     1   NaN  21.0    31    41  51.0    61    71  81.0     A
1       row2   200   NaN  22.0    32    42  52.

In [ ]:
#while reading data ,data might have some errors,dulpicate values,missing values,formating error
#panda provide solution such data


#===finding missing data===

df.isna()

#==5 column and 15 rows it will give true for missing value
================Or=======
by itteration

for i in range(len(df)):
    for col in df.columns():
    if df.loc[i,col].isna():
        .....
        
=======want to fill missing value====
df.filna(0)

#want to delete the row that have missing value

df.dropna()

#will drop all the rows that have missing value

df["height"].fillna[x] #fill the missing data

# ===========Pandas - Cleaning Empty Cells=============

In [26]:
d=pd.read_csv("demo.csv",sep="\t")
print(d)

  Unnamed: 0  col1  col2  col3  col4  col5  col6  col7  col8  col9 col10
0       row1     1   NaN    21    31    41    51    61    71    81     A
1       row2   200   NaN    22    32    42    52    62    72    82     B
2       row3     3  13.0    23    33    43    53    63    73    83     C
3       row4     0  14.0    24     0    44    54    64    74    84     D
4       row5     5  15.0    25    35    45    55    65    75    85     E
5       row6     6  16.0    26    36    46    56    66    76    86     F
6       row7     7  17.0    27    37    47    57    67    77    87     G
7       row8     8  18.0    28    38    48    58    68    78    88     H
8       row9     9  19.0    29    39    49    59    69    79    89     I
9      row10    10  20.0    30    40    50    60    70    80    90     J


# ============= isna()==========

# Checking for null/nan value

In [15]:
d.isna()

Unnamed: 0   col1   col2   col3   col4   col5   col6   col7   col8   col9  \
0       False  False   True  False  False  False  False  False  False  False   
1       False  False   True  False  False  False  False  False  False  False   
2       False  False  False  False  False  False  False  False  False  False   
3       False  False  False  False  False  False  False  False  False  False   
4       False  False  False  False  False  False  False  False  False  False   
5       False  False  False  False  False  False  False  False  False  False   
6       False  False  False  False  False  False  False  False  False  False   
7       False  False  False  False  False  False  False  False  False  False   
8       False  False  False  False  False  False  False  False  False  False   
9       False  False  False  False  False  False  False  False  False  False   

   col10  
0  False  
1  False  
2  False  
3  False  
4  False  
5  False  
6  False  
7  False  
8  False  
9  False

In [22]:
pd.isna(d['col2'])

0     True
1     True
2    False
3    False
4    False
5    False
6    False
7    False
8    False
9    False
Name: col2, dtype: bool

In [25]:
pd.isna(d['col3'])

0    False
1    False
2    False
3    False
4    False
5    False
6    False
7    False
8    False
9    False
Name: col3, dtype: bool

In [24]:
pd.isna(d['col3']).sum()

0

# To see how many nan value there in each column

In [16]:
d.isna().sum()

Unnamed: 0    0
col1          0
col2          2
col3          0
col4          0
col5          0
col6          0
col7          0
col8          0
col9          0
col10         0
dtype: int64

# ================Remove Rows==========

### >>One way to deal with empty cells is to remove rows that contain empty cells.

# ================dropna()=================

# to delete the row that have missing value

### Note: By default, the dropna() method returns a new DataFrame, and will not change the original.

### If you want to change the original DataFrame, use the ( inplace = True ) argument:

In [27]:
d.dropna()

Unnamed: 0  col1  col2  col3  col4  col5  col6  col7  col8  col9 col10
2       row3     3  13.0    23    33    43    53    63    73    83     C
3       row4     0  14.0    24     0    44    54    64    74    84     D
4       row5     5  15.0    25    35    45    55    65    75    85     E
5       row6     6  16.0    26    36    46    56    66    76    86     F
6       row7     7  17.0    27    37    47    57    67    77    87     G
7       row8     8  18.0    28    38    48    58    68    78    88     H
8       row9     9  19.0    29    39    49    59    69    79    89     I
9      row10    10  20.0    30    40    50    60    70    80    90     J

In [28]:
print(d)  #not changing the original dataframe

  Unnamed: 0  col1  col2  col3  col4  col5  col6  col7  col8  col9 col10
0       row1     1   NaN    21    31    41    51    61    71    81     A
1       row2   200   NaN    22    32    42    52    62    72    82     B
2       row3     3  13.0    23    33    43    53    63    73    83     C
3       row4     0  14.0    24     0    44    54    64    74    84     D
4       row5     5  15.0    25    35    45    55    65    75    85     E
5       row6     6  16.0    26    36    46    56    66    76    86     F
6       row7     7  17.0    27    37    47    57    67    77    87     G
7       row8     8  18.0    28    38    48    58    68    78    88     H
8       row9     9  19.0    29    39    49    59    69    79    89     I
9      row10    10  20.0    30    40    50    60    70    80    90     J


## Note: Now, the dropna(inplace = True) will NOT return a new DataFrame, 
## but it will remove all rows containg NULL values from the original DataFrame.

In [29]:
d.dropna(inplace=True) 

In [30]:
print(d)   #actually changing the original dataframe

  Unnamed: 0  col1  col2  col3  col4  col5  col6  col7  col8  col9 col10
2       row3     3  13.0    23    33    43    53    63    73    83     C
3       row4     0  14.0    24     0    44    54    64    74    84     D
4       row5     5  15.0    25    35    45    55    65    75    85     E
5       row6     6  16.0    26    36    46    56    66    76    86     F
6       row7     7  17.0    27    37    47    57    67    77    87     G
7       row8     8  18.0    28    38    48    58    68    78    88     H
8       row9     9  19.0    29    39    49    59    69    79    89     I
9      row10    10  20.0    30    40    50    60    70    80    90     J


In [17]:
#storing the dataframe that have deleted missing value row into another variable x

x=d.dropna()



In [18]:
print(x)

  Unnamed: 0  col1  col2  col3  col4  col5  col6  col7  col8  col9 col10
2       row3     3  13.0    23    33    43    53    63    73    83     C
3       row4     0  14.0    24     0    44    54    64    74    84     D
4       row5     5  15.0    25    35    45    55    65    75    85     E
5       row6     6  16.0    26    36    46    56    66    76    86     F
6       row7     7  17.0    27    37    47    57    67    77    87     G
7       row8     8  18.0    28    38    48    58    68    78    88     H
8       row9     9  19.0    29    39    49    59    69    79    89     I
9      row10    10  20.0    30    40    50    60    70    80    90     J


In [ ]:
## HAndling Missing Values

## Drop nan values

# df.dropna(axis=0)

# df.dropna(axis=1)


# ==========fillna()===========

# ==========Replace Empty Values===========

### Another way of dealing with empty cells is to insert a new value instead.

### This way you do not have to delete entire rows just because of some empty cells.

### The fillna() method allows us to replace empty cells with a value:

In [32]:
d=pd.read_csv("demo.csv",sep="\t")
print(d)

  Unnamed: 0  col1  col2  col3  col4  col5  col6  col7  col8  col9 col10
0       row1     1   NaN  21.0    31    41  51.0    61    71  81.0     A
1       row2   200   NaN  22.0    32    42  52.0    62    72  82.0     B
2       row3     3  13.0  23.0    33    43  53.0    63    73   NaN     C
3       row4     0  14.0  24.0     0    44  54.0    64    74  84.0     D
4       row5     5  15.0  25.0    35    45  55.0    65    75  85.0     E
5       row6     6  16.0  26.0    36    46  56.0    66    76  86.0     F
6       row7     7  17.0  27.0    37    47  57.0    67    77  87.0     G
7       row8     8  18.0  28.0    38    48  58.0    68    78  88.0     H
8       row9     9  19.0   NaN    39    49   NaN    69    79  89.0     I
9      row10    10  20.0  30.0    40    50  60.0    70    80  90.0     J


### =======Replace Only For a Specified Columns=========

### To only replace empty values for one column, specify the column name for the DataFrame:

In [34]:
d['col6'].fillna(1000,inplace=True)

In [35]:
print(d)  #column 6's NAN value has been changed with 1000

  Unnamed: 0  col1  col2  col3  col4  col5    col6  col7  col8  col9 col10
0       row1     1   NaN  21.0    31    41    51.0    61    71  81.0     A
1       row2   200   NaN  22.0    32    42    52.0    62    72  82.0     B
2       row3     3  13.0  23.0    33    43    53.0    63    73   NaN     C
3       row4     0  14.0  24.0     0    44    54.0    64    74  84.0     D
4       row5     5  15.0  25.0    35    45    55.0    65    75  85.0     E
5       row6     6  16.0  26.0    36    46    56.0    66    76  86.0     F
6       row7     7  17.0  27.0    37    47    57.0    67    77  87.0     G
7       row8     8  18.0  28.0    38    48    58.0    68    78  88.0     H
8       row9     9  19.0   NaN    39    49  1000.0    69    79  89.0     I
9      row10    10  20.0  30.0    40    50    60.0    70    80  90.0     J


# ======Replaces all empty cells in the whole Data Frame=====

### Replace NULL values with the number 2000 :

In [36]:
d.fillna(2000,inplace=True)

In [37]:
print(d)   #here all null has been changed to 2000

  Unnamed: 0  col1    col2    col3  col4  col5    col6  col7  col8    col9  \
0       row1     1  2000.0    21.0    31    41    51.0    61    71    81.0   
1       row2   200  2000.0    22.0    32    42    52.0    62    72    82.0   
2       row3     3    13.0    23.0    33    43    53.0    63    73  2000.0   
3       row4     0    14.0    24.0     0    44    54.0    64    74    84.0   
4       row5     5    15.0    25.0    35    45    55.0    65    75    85.0   
5       row6     6    16.0    26.0    36    46    56.0    66    76    86.0   
6       row7     7    17.0    27.0    37    47    57.0    67    77    87.0   
7       row8     8    18.0    28.0    38    48    58.0    68    78    88.0   
8       row9     9    19.0  2000.0    39    49  1000.0    69    79    89.0   
9      row10    10    20.0    30.0    40    50    60.0    70    80    90.0   

  col10  
0     A  
1     B  
2     C  
3     D  
4     E  
5     F  
6     G  
7     H  
8     I  
9     J  


# ====Replace Using Mean, Median, or Mode====

### A common way to replace empty cells, is to calculate the mean, median or mode value of the column.

### Pandas uses the mean() median() and mode() methods to calculate the respective values for a specified column:

# Mean = the average value (the sum of all values divided by number of values).

# Median = the value in the middle, after you have sorted all values ascending.

# Mode = the value that appears most frequently.

In [40]:
#  Calculate the MEAN, and replace any empty values with it:


df=d=pd.read_csv("demo.csv",sep="\t")
print(df,"\n","\n")

x = df["col2"].mean()

df["col2"].fillna(x, inplace = True) #nan value has been replaced with mean of column2 i.e 16.5

print(df,"\n","\n")

print(x)

  Unnamed: 0  col1  col2  col3  col4  col5  col6  col7  col8  col9 col10
0       row1     1   NaN  21.0    31    41  51.0    61    71  81.0     A
1       row2   200   NaN  22.0    32    42  52.0    62    72  82.0     B
2       row3     3  13.0  23.0    33    43  53.0    63    73   NaN     C
3       row4     0  14.0  24.0     0    44  54.0    64    74  84.0     D
4       row5     5  15.0  25.0    35    45  55.0    65    75  85.0     E
5       row6     6  16.0  26.0    36    46  56.0    66    76  86.0     F
6       row7     7  17.0  27.0    37    47  57.0    67    77  87.0     G
7       row8     8  18.0  28.0    38    48  58.0    68    78  88.0     H
8       row9     9  19.0   NaN    39    49   NaN    69    79  89.0     I
9      row10    10  20.0  30.0    40    50  60.0    70    80  90.0     J 
 

  Unnamed: 0  col1  col2  col3  col4  col5  col6  col7  col8  col9 col10
0       row1     1  16.5  21.0    31    41  51.0    61    71  81.0     A
1       row2   200  16.5  22.0    32    42  52.

In [42]:
# Calculate the MEDIAN, and replace any empty values with it:

df = pd.read_csv("demo.csv",sep="\t")

print(df,"\n","\n")

x = df["col3"].median()

df["col3"].fillna(x, inplace = True)


print(x,"\n","\n")
print(df,"\n","\n")

  Unnamed: 0  col1  col2  col3  col4  col5  col6  col7  col8  col9 col10
0       row1     1   NaN  21.0    31    41  51.0    61    71  81.0     A
1       row2   200   NaN  22.0    32    42  52.0    62    72  82.0     B
2       row3     3  13.0  23.0    33    43  53.0    63    73   NaN     C
3       row4     0  14.0  24.0     0    44  54.0    64    74  84.0     D
4       row5     5  15.0  25.0    35    45  55.0    65    75  85.0     E
5       row6     6  16.0  26.0    36    46  56.0    66    76  86.0     F
6       row7     7  17.0  27.0    37    47  57.0    67    77  87.0     G
7       row8     8  18.0  28.0    38    48  58.0    68    78  88.0     H
8       row9     9  19.0   NaN    39    49   NaN    69    79  89.0     I
9      row10    10  20.0  30.0    40    50  60.0    70    80  90.0     J 
 

25.0 
 

  Unnamed: 0  col1  col2  col3  col4  col5  col6  col7  col8  col9 col10
0       row1     1   NaN  21.0    31    41  51.0    61    71  81.0     A
1       row2   200   NaN  22.0    32  

In [43]:
# Calculate the MODE, and replace any empty values with it:


df = pd.read_csv("demo.csv",sep="\t")

print(df,"\n","\n")

x = df["col6"].mode()[0]

df["col6"].fillna(x, inplace = True)

print(x,"\n","\n")
print(df,"\n","\n")

  Unnamed: 0  col1  col2  col3  col4  col5  col6  col7  col8  col9 col10
0       row1     1   NaN  21.0    31    41  51.0    61    71  81.0     A
1       row2   200   NaN  22.0    32    42  52.0    62    72  82.0     B
2       row3     3  13.0  23.0    33    43  53.0    63    73   NaN     C
3       row4     0  14.0  24.0     0    44  54.0    64    74  84.0     D
4       row5     5  15.0  25.0    35    45  55.0    65    75  85.0     E
5       row6     6  16.0  26.0    36    46  56.0    66    76  86.0     F
6       row7     7  17.0  27.0    37    47  57.0    67    77  87.0     G
7       row8     8  18.0  28.0    38    48  58.0    68    78  88.0     H
8       row9     9  19.0   NaN    39    49   NaN    69    79  89.0     I
9      row10    10  20.0  30.0    40    50  60.0    70    80  90.0     J 
 

51.0 
 

  Unnamed: 0  col1  col2  col3  col4  col5  col6  col7  col8  col9 col10
0       row1     1   NaN  21.0    31    41  51.0    61    71  81.0     A
1       row2   200   NaN  22.0    32  

# =====Pandas - Cleaning Data of Wrong Format=====

###  ===Data of Wrong Format===

### Cells with data of wrong format can make it difficult, or even impossible, to analyze data.

### To fix it, you have two options: remove the rows, or convert all cells in the columns into the same format.

# Convert Into a Correct Format

# ====Pandas has a to_datetime() method ====

In [46]:


df = pd.read_csv('dirtydata.csv')

print(df,'\n','\n')

df['Date'] = pd.to_datetime(df['Date'])

print(df)


    Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'  

# ==Removing Rows==

### The result from the converting in the example above gave us a NaT value, 
### which can be handled as a NULL value, and we can remove the row by using the dropna() method.

In [47]:
df.dropna(inplace=True)
print(df)

    Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7        450 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
19        60 2020-12-19    103 

# Pandas - Fixing Wrong Data

# Wrong Data

### "Wrong data" does not have to be "empty cells" or "wrong format", it can just be wrong, like if someone registered "199" instead of "1.99".

### Sometimes you can spot wrong data by looking at the data set, because you have an expectation of what it should be.

### If you take a look at our data set, you can see that in row 7, the duration is 450, but for all the other rows the duration is between 30 and 60.

### It doesn't have to be wrong, but taking in consideration that this is the data set of someone's workout sessions, we conclude with the fact that this person did not work out in 450 minutes.

# ======Replacing Values======

### One way to fix wrong values is to replace them with something else.

In [52]:


df.loc[7,"Duration"]=50

print(df,'\n','\n')

    Duration       Date  Pulse  Maxpulse  Calories  (7, Duration)
0         60 2020-12-01    110       130     409.1             50
1         60 2020-12-02    117       145     479.0             50
2         60 2020-12-03    103       135     340.0             50
3         45 2020-12-04    109       175     282.4             50
4         45 2020-12-05    117       148     406.0             50
5         60 2020-12-06    102       127     300.0             50
6         60 2020-12-07    110       136     374.0             50
7         50 2020-12-08    104       134     253.3             50
8         30 2020-12-09    109       133     195.1             50
9         60 2020-12-10     98       124     269.0             50
10        60 2020-12-11    103       147     329.3             50
11        60 2020-12-12    100       120     250.7             50
12        60 2020-12-12    100       120     250.7             50
13        60 2020-12-13    106       128     345.3             50
14        

For small data sets you might be able to replace the wrong data one by one, but not for big data sets.

To replace wrong data for larger data sets you can create some rules,
e.g. set some boundaries for legal values, and replace any values that are outside of the boundaries.

In [55]:
print(df)

    Duration       Date  Pulse  Maxpulse  Calories  (7, Duration)
0         60 2020-12-01    110       130     409.1             50
1         60 2020-12-02    117       145     479.0             50
2         60 2020-12-03    103       135     340.0             50
3         45 2020-12-04    109       175     282.4             50
4         45 2020-12-05    117       148     406.0             50
5         60 2020-12-06    102       127     300.0             50
6         60 2020-12-07    110       136     374.0             50
7         50 2020-12-08    104       134     253.3             50
8         30 2020-12-09    109       133     195.1             50
9         60 2020-12-10     98       124     269.0             50
10        60 2020-12-11    103       147     329.3             50
11        60 2020-12-12    100       120     250.7             50
12        60 2020-12-12    100       120     250.7             50
13        60 2020-12-13    106       128     345.3             50
14        

In [56]:
print(df.index)

Int64Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
            17, 19, 20, 21, 23, 24, 25, 26, 27, 29, 30, 31],
           dtype='int64')


In [57]:
for i in df.index:
    if df.loc[i,"Duration"]>50:
        df.loc[i,"Duration"]=100
        
print(df)
        

    Duration       Date  Pulse  Maxpulse  Calories  (7, Duration)
0        100 2020-12-01    110       130     409.1             50
1        100 2020-12-02    117       145     479.0             50
2        100 2020-12-03    103       135     340.0             50
3         45 2020-12-04    109       175     282.4             50
4         45 2020-12-05    117       148     406.0             50
5        100 2020-12-06    102       127     300.0             50
6        100 2020-12-07    110       136     374.0             50
7         50 2020-12-08    104       134     253.3             50
8         30 2020-12-09    109       133     195.1             50
9        100 2020-12-10     98       124     269.0             50
10       100 2020-12-11    103       147     329.3             50
11       100 2020-12-12    100       120     250.7             50
12       100 2020-12-12    100       120     250.7             50
13       100 2020-12-13    106       128     345.3             50
14       1

# ====Removing Rows====


### Another way of handling wrong data is to remove the rows that contains wrong data.

### This way you do not have to find out what to replace them with, and there is a good chance you do not need them to do your analyses.

In [61]:
for i in df.index:
    if df.loc[i,'Duration']<50:
        df.drop(i,inplace = True)
        
        
print(df)

    Duration       Date  Pulse  Maxpulse  Calories  (7, Duration)
0        100 2020-12-01    110       130     409.1             50
1        100 2020-12-02    117       145     479.0             50
2        100 2020-12-03    103       135     340.0             50
5        100 2020-12-06    102       127     300.0             50
6        100 2020-12-07    110       136     374.0             50
7         50 2020-12-08    104       134     253.3             50
9        100 2020-12-10     98       124     269.0             50
10       100 2020-12-11    103       147     329.3             50
11       100 2020-12-12    100       120     250.7             50
12       100 2020-12-12    100       120     250.7             50
13       100 2020-12-13    106       128     345.3             50
14       100 2020-12-14    104       132     379.3             50
15       100 2020-12-15     98       123     275.0             50
16       100 2020-12-16     98       120     215.2             50
17       1

# ==========duplicated() method ========

### Returns True for every row that is a duplicate, othwerwise False:

# Pandas - Removing Duplicates

# ==Discovering Duplicates==

### Duplicate rows are rows that have been registered more than one time.

In [66]:
df = pd.read_csv('dirtydata.csv')

print(df,'\n','\n')

    Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'  

By taking a look at our test data set, we can assume that row 11 and 12 are duplicates.

To discover duplicates, we can use the duplicated() method.

The duplicated() method returns a Boolean values for each row:

In [67]:
print(df.duplicated())

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12     True
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20    False
21    False
22    False
23    False
24    False
25    False
26    False
27    False
28    False
29    False
30    False
31    False
dtype: bool


# ==================drop_duplicates() method ==========

# ======Removing Duplicates======

### To remove duplicates, use the drop_duplicates() method.

In [68]:
# Remove all duplicates:

df.drop_duplicates(inplace = True)

print(df) #12 row deleted

    Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'    100       120     300.0
18        45  '2020/12/18'  

Remember: The (inplace = True) will make sure that the method does NOT return a new DataFrame, but it will remove all duplicates from the original DataFrame.

# ====inplace====

## amra jai e kori dataframe fillna,dropna,drop_duplicate....ai gulor result show korbe kintu original dataframe e change hobe na just oi kaj gulo kore print korbe....original dataframe e change korte gyle hoy onno akta dataframe create kore result ta store kore rakte hobe(ex:df2=df.drop_duplicates) kintu arom korte gyle onek dataframe create kora hoye jabe tai amra (inplace=True) use kore korbo taholai original dataframe e change hobe noy to hobe na

In [3]:
df = pd.read_csv('dirtydata.csv')

df.head()

Duration          Date  Pulse  Maxpulse  Calories
0        60  '2020/12/01'    110       130     409.1
1        60  '2020/12/02'    117       145     479.0
2        60  '2020/12/03'    103       135     340.0
3        45  '2020/12/04'    109       175     282.4
4        45  '2020/12/05'    117       148     406.0

In [6]:
x=df[df['Duration']==60]

In [7]:
x

Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'    100       120     300.0
19        60  '2020/12/19'    103       123     323.0
21        60  '2020/12/21'    108       131     364.2
23        60  '2020/12/23'    130       101     300.0
25        60  '2020/12/25'    102       126     334.5
26        60      20201226    100       120     250.0
27        60  '2020/12/27'     92       118     241.0
28        60  '2020/12/28'    103       132       NaN
29        60  '2020/12/29'    100       132     280.0
30        60  '2020/12/30'    102       129     380.3
31        60  '2020/12/31'     92       115     243.0

In [8]:
x=df[df['Duration']==60].shape[0]

In [9]:
x

24